python/grass/jupyter/timeseries.py

```diff
@@ -0,0 +1,383 @@
+
+        # Handle Regions
+        region_manager = RegionManagerForTimeSeries(use_region, saved_region, self._env)
+        region_manager.set_region_from_timeseries(self.timeseries)
+
+    @property
+    def overlay(self):
+        """Add overlay to TimeSeries visualization"""
+        self._layers_rendered = False
+        return self._overlays
+
+    @property
+    def baselayer(self):
+        """Add base layer to TimeSeries visualization"""
+        self._layers_rendered = False
+        return self._baselayers
+
+    def set_background_color(self, color):
+        """Set background color of images.
+
+        Passed to d.rast and d.erase. Either a standard color name or R:G:B triplet.
+        Default is White."""
+        self._bgcolor = color
+        self._layers_rendered = False
+
+    def d_legend(self, **kwargs):
+        """Wrapper for d.legend. Passes keyword arguments to d.legend in render_layers
+        method.
+        """
+        self._legend = kwargs
+        # If d_legend has been called, we need to re-render layers
+        self._layers_rendered = False
+
+    def _render_legend(self, img):
+        """Add legend to GrassRenderer instance"""
+        info = gs.parse_command(
+            "t.info", input=self.timeseries, flags="g", env=self._env
+        )
+        min_min = info["min_min"]
+        max_max = info["max_max"]
+        img.d_legend(
+            raster=self._layers[0],
+            range=f"{min_min}, {max_max}",
+            **self._legend,
+        )
+
+    def _render_blank_layer(self, filename):
+        """Write blank image for gaps in time series"""
+        # Render image
+        img = GrassRenderer(filename=filename, use_region=True, env=self._env)
+        # Write blank image
+        img.d_erase(bgcolor=self._bgcolor)
+        # Add legend if needed
+        if self._legend:
+            self._render_legend(img)
+
+    def render(self):
+        """Renders map for each time-step in space-time dataset and save to PNG
+        image in temporary directory.
+
+        Must be run before creating a visualization (i.e. time_slider or animate).
+
+        Can be time-consuming to run with large space-time datasets.
+        """
+
+        # Render each layer
+        for date, layer in self._date_layer_dict.items():
+            if layer == "None":
+                filename = os.path.join(self._tmpdir.name, "None.png")
+                self._date_filename_dict[date] = filename
+                self._render_blank_layer(filename)
+            else:
+                # Create image file
+                filename = os.path.join(self._tmpdir.name, f"{layer}.png")
+                self._date_filename_dict[date] = filename
+                # Render image
+                img = GrassRenderer(filename=filename, use_region=True, env=self._env)
+                # Fill image background
+                img.d_erase(bgcolor=self._bgcolor)
+                for grass_module, kwargs in self._baselayers.calls:
```

Comment:
An idea for rendering the background just once: Specify the image to render for a separate instance of GrassRenderer (base layers + background color). Then use copy that image and use copies for the time series renderers, but tell them not to draw the background color (no d.erase), i.e., use the image as is.

The blank layer also needs to be rendered just once. It seems you are reusing the name already, so there is in fact just one file. It's just rendered more that once now.